Implement `createCriterion` and `getUserSettings` calls

domienderle opened the pull request:
Closes #110 #95

## What I have made

<!-- write down what changes have been made, what was removed/added/changed -->
<!-- e.g. I added a service, which does x -->

First of all I created the `UserSettings` in the database. Then I implemented the method `getUserSettings` to be able to retrieve the user settings.

I adjusted the methods to create a criterion on the repository, service and validation layer to distinguish between user and project criteria. This includes the following behaviour:
- everyone can create a user criterion (no restrictions)
- as soon as a project is created, the default criteria of the users `UserSettings` are duplicated and stored as project criteria
- every project admin is allowed to add new project criteria

**Note:** This branch was rebased onto the `updateCriterion` branch, which is why only the last four commits are relevant for this PR.
 
## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does
not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

domienderle committed:
feat: adjust repository layer for createCriterion call

domienderle committed:
feat: adjust service layer for createCriterion call

domienderle committed:
feat: adjust validation layer for createCriterion call

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:
```diff
@@ -28,6 +40,21 @@ interface ICriterionTableRepo {
      */
     suspend fun createCriterion(request: CriterionOuterClass.Criterion.Create, userId: UUID): Criterion
+
+    suspend fun getAllUserCriteria(userId: UUID): List<Criterion>
```
Provide a comment for this method.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:
```diff
@@ -43,15 +70,30 @@ interface ICriterionTableRepo {
+            if (!request.projectId.isNullOrEmpty()) {
+                val projectUUID = parseUUID(request.projectId, EntityType.PROJECT)
+                projectEntityId = getProjectEntityId(projectUUID)
+            }
```
The `getUserEntityId` and `getProjectEntityId` calls are obsolete. I initially thought we would require the `EntityID` object, but we can simply use the wrapped value. Additionally, we only need to check for nullability because we assume that the projectId is not empty.

In the end we can reduce this to a single line.

```suggestion
        db.dbQuery {
            val projectId = request.projectId?.let { parseUUID(it, EntityType.PROJECT) }
```

Further down, we just have to adjust the used IDs:
```kotlin
it[CriterionTable.projectId] = projectId
it[createdBy] = userId
```

domienderle replied:
I don't think this is correct. I tried it a couple times now and I get an referential integrity error if remove the `getProjectEntityId` or the `getUserEntitiyId`. I did not find any solution without those that worked.

domienderle replied:
The second point is with the not null check. I don't think this is right either. As far as I know if the `project_id` field of an grpc call is not defined it is set empty. This behaviour can also be seen in the tests, if we don't set the `project_id`. Therefore we can only remove the nullability check but not the emptiness check.

Slartibartfass2 replied:
Interesting, I thought we had not used the entityId references in other repo methods, but apparently we didn't.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -120,23 +129,48 @@ class ProjectTableRepo(
+                        .build(),
+                    userId,
+                ),
+            )
```
Let's apply this logic in the service layer. The repository layer is designed to only execute the most basic CRUD operations. By coupling two repositories, we increase the complexity, which is not necessary, and so far we haven't needed to. Also, we don't use the criteria return value, so there's no need for returning it.

domienderle replied:
I removed this from the repository layer. Since we discussed about how to create criteria if a project is created, I also added the user settings and the criteria are now created depending on the users default criteria stored in the user settings.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepo.kt:
```diff
@@ -120,23 +129,48 @@ class ProjectTableRepo(
-            it[fetcherApis] = emptyList()
-            it[createdBy] = userEntityId
+        val project = db.dbQuery {
+            val userEntityId = getUserEntityId(userId)
```
As already mentioned in the `CriterionTableRepo`, we don't need to use `getUserEntityId` anymore since I made a wrong assumption. Please remove this for me.

domienderle replied:
See above.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -2,30 +2,181 @@ package se.uulm.snowballr.backend.service
+                    EntityType.PROJECT,
+                    projectId.toString(),
+                    AccessType.READ,
+                    currentUser.id.toString(),
```
```suggestion
                    it,
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -2,30 +2,181 @@ package se.uulm.snowballr.backend.service
+                EntityType.CRITERION,
+                criterion.id.toString(),
+                accessType,
+                currentUser.id.toString(),
```
```suggestion
                it,
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -2,30 +2,181 @@ package se.uulm.snowballr.backend.service
+
+    override suspend fun createCriterion(request: CriterionOuterClass.Criterion.Create): CriterionOuterClass.Criterion {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+        if (!request.projectId.isNullOrEmpty()) {
```
Only check for non-null project IDs since we assume that the project ID is either valid or null.

domienderle replied:
I removed the null check.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/CriterionValidator.kt:
```diff
@@ -10,27 +11,87 @@ const val CRITERION_TAG_MAX_LENGTH = 10
             {
-                ensureFieldNonBlank("tag", request.tag)
-                ensureFieldLength("tag", request.tag, CRITERION_TAG_MAX_LENGTH)
+                if (!request.projectId.isNullOrEmpty()) {
```
Only check for non-nullability. This should fail if the projectId is empty.

domienderle replied:
I removed the null check.

Slartibartfass2 reviewed (changes requested):
This is it for now. I wait to review the tests until after the rebase and after we agree on where to put the user criteria copy logic.

Good job so far. I see that your implementations are getting  cleaner every time.

domienderle committed:
feat: add userSettings to the database

domienderle committed:
feat: add repository layer for getUserSettings

domienderle committed:
feat: add service layer for getUserSettings

domienderle committed:
feat: add grpc layer for getUserSettings

domienderle commented:
I have added your review. Additionally, I have changed the functionality of creating criteria if a project is newly created. Therefore I have added the user settings. If a project gets created now, the default criteria of the user settings are duplicated as project criteria.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/auth/DummyUser.kt:
```diff
@@ -20,4 +21,13 @@ object DummyUser {
+
+    var showHotkeys: Boolean = true
+    var reviewMode: Boolean = false
+    var criteriaIds: List<UUID> = emptyList()
+    var similarityThreshold: Float = 0F
+    var decisionMatrix: ByteArray = ProjectOuterClass.ReviewDecisionMatrix.getDefaultInstance().toByteArray()
+    var fetcherApis: List<String> = emptyList()
+    var snowballingType: ProjectOuterClass.SnowballingType = ProjectOuterClass.SnowballingType.SNOWBALLING_TYPE_BOTH
+    var reviewMaybeAllowed: Boolean = true
```
When we designed the mockups, we decided that only the criteria and the fetchers are stored in the user settings and can be used as defaults for project creation. Did we change that since?

Storing simple values like the threshold, the snowballing type, and whether maybe is allowed seems obsolete as it is not much work to define it after the project was created. With the defaults for fetchers and criteria, we thought that this would reduce the time required to setup a project since many criteria and especially fetchers may be duplicated across several projects.

Slartibartfass2 replied:
I just saw that this is part of the API spec. We will discuss this in our team channel.

domienderle replied:
Since, we had now clear solution in the team channel I would propose to keep the extra settings. I know it is not a lot of effort add this later on, but imo it is not a problem to have this from the beginning since it neither affects the performance a lot nor wastes a lot of storage. But I am also not aware of common standards in such situations and if you think we should remove them to keep our system as clean as possible it is also ok for me.

Slartibartfass2 replied:
Let's keep them. We can also remove them later if we decide differently.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/auth/DummyUser.kt:
```diff
@@ -20,4 +21,13 @@ object DummyUser {
     var passwordHash: String = PasswordUtils.hashPassword(password)
     var role: UserOuterClass.UserRole = UserOuterClass.UserRole.USER_ROLE_ADMIN
     var status: UserOuterClass.UserStatus = UserOuterClass.UserStatus.USER_STATUS_ACTIVE
+
```
Either remove the empty line or put a comment there
```suggestion
    // user settings
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/validation/CriterionValidator.kt:
```diff
@@ -29,7 +29,11 @@ private const val MASK_CATEGORY = "criterion.category"
         zipOrAccumulate(
-            { ensureIdValidity(FIELD_PROJECT_ID, request.projectId) },
+            {
+                if (!request.projectId.isEmpty()) {
```
```suggestion
                if (request.projectId.isNotEmpty()) {
```

Slartibartfass2 replied:
Also, add a test case for this

domienderle replied:
Good call. I forgot this. I added.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -76,35 +76,31 @@ class CriterionService(
         val members = when (accessType) {
             AccessType.READ -> projectMemberRepo.getMembersOfProject(project.id)
+            AccessType.CREATE -> projectMemberRepo.getAllProjectAdmins(project.id)
             AccessType.UPDATE -> projectMemberRepo.getAllProjectAdmins(project.id)
```
We combine those two cases.

```suggestion
            AccessType.UPDATE, AccessType.CREATE -> projectMemberRepo.getAllProjectAdmins(project.id)
```

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -76,35 +76,31 @@ class CriterionService(
                 throw UnauthorizedException.Single(
-                    EntityType.CRITERION,
-                    criterionId,
+                    EntityType.PROJECT,
+                    projectId.toString(),
                     AccessType.READ,
-                    currentUser.id.toString(),
+                    it,
                 )
```
This would throw an exception with the message "User with ID '...' is not authorized to read project with ID '...'.". Is this correct? Wouldn't we want something like "not authorized to create a criterion for project with ID '...'"?

domienderle replied:
Yes, you are completely right. I also thought about this. The problem I faced was, that we would have to distinguish between `READ`, `CREATE` and `UPDATE`. For read and update we could use the already defined exception with the already defined criterion id. For create we would have to create a new exception and formulate something like you said. This would mean a new exception and a new branching for this exception. I implemented an example how this could look like. I am not happy with this, but I don't think we have a better option if we want to have the exception like this. What do you think?

I know that I have to adjust the tests and the comments as well. But I first want to have your opinion about it before I start with the tests.

Slartibartfass2 replied:
I agree this is quite complicated, and our current system doesn't support this. Let's undo it and find a better solution in another issue. I created #229 for that. Please take a look and add further information if you have some.

domienderle replied:
All right. Sounds good.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -76,35 +76,31 @@ class CriterionService(
             verifyServerAdminRole(currentUser) {
                 throw SnowballRException.FailedPreconditionException(
-                    "The project is not active. Cannot update criterion with the ID: $criterionId",
+                    "The project is not active. Cannot create or update criteria.",
```
Let's keep the criterion ID included so that we can track the exception. Or at least the project ID.

domienderle replied:
I also added branching here. Let me hear your opinion about this as well.

Slartibartfass2 replied:
We can only set the message according to the state and then throw the exception with the according message:
```kotlin
            verifyServerAdminRole(currentUser) {
                val message = if (accessType == AccessType.CREATE) {
                    "Cannot create a criterion for the project with the id: $projectId."
                } else {
                    "Cannot update criterion with the ID: ${criterion!!.id}"
                }
                throw SnowballRException.FailedPreconditionException("The project is not active. $message")
            }
```

I think this is fine for now but probably also requires a clean refactoring later.

domienderle replied:
That is a good idea. Thank you.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/CriterionService.kt:
```diff
@@ -163,8 +159,17 @@ class CriterionService(
-        repo.createCriterion(request, GrpcContext.getUserIdFromContext()).toGrpcCriterion()
+    override suspend fun createCriterion(request: CriterionOuterClass.Criterion.Create): CriterionOuterClass.Criterion {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+        if (!request.projectId.isEmpty()) {
```
```suggestion
        if (request.projectId.isNotEmpty()) {
```

Do you use the SonarQube plugin for IntelliJ?

domienderle replied:
No, not since I switched from the wsl Intellij to normal Intellij. Thanks for the reminder.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -12,17 +12,39 @@ import org.junit.jupiter.api.assertThrows
-    fun setUpTest() {
+    fun setupTest() {
         every { GrpcContext.getUserIdFromContext() } throws NotImplementedError()
+        coEvery { userRepoMock.getUserById(GrpcContext.getUserIdFromContext()) } throws NotImplementedError()
```
The other call is mocked one call above.
```suggestion
        coEvery { userRepoMock.getUserById(any()) } throws NotImplementedError()
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -45,13 +67,72 @@ class CreateCriterionTest : MainServiceTest() {
-    fun `When a criterion is correctly created, then no exception is thrown`() = runTest {
-        val request = CriterionOuterClass.Criterion.Create.getDefaultInstance()
-        val criterion = DataBuilder.createExampleCriterion()
+    fun `When an admin user creates a project criterion, then an unauthorized exception is thrown`() = runTest {
```
The name says it throws, but the assertion says it doesn't.

domienderle replied:
Good catch.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -45,13 +67,72 @@ class CreateCriterionTest : MainServiceTest() {
+    }
+
+    @Test
+    fun `When a user, who is project admin creates a project criterion, then no exception is thrown`() = runTest {
```
```suggestion
    fun `When a project admin creates a project criterion, then no exception is thrown`() = runTest {
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -45,13 +67,72 @@ class CreateCriterionTest : MainServiceTest() {
     }
+
+    @Test
+    fun `When a user, who is not project admin creates a project criterion, then an unauthorized exception is thrown`() =
```
```suggestion
    fun `When a non project admin creates a project criterion, then an unauthorized exception is thrown`() =
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -45,13 +67,72 @@ class CreateCriterionTest : MainServiceTest() {
+
+            assertThrows<SnowballRException.UnauthorizedException.Single> { mainService.createCriterion(request) }
+        }
+
```
Also add test cases for:
- creating a project criterion for an inactive project
- creating a user criterion

domienderle replied:
All right.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -12,17 +12,39 @@ import org.junit.jupiter.api.assertThrows
+        coEvery { projectMemberRepoMock.addUserToProject(any(), any()) } throws NotImplementedError()
+        coEvery {
+            projectMemberRepoMock.promoteProjectMemberToAdmin(any(), any())
+        } throws NotImplementedError()
```
Both these calls are not used.

domienderle replied:
I have removed them.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:
```diff
@@ -84,13 +94,12 @@ class CriterionTableRepo(
-            // Get project reference
-            val projectEntityId = getProjectEntityId(projectUUID)
+            var projectEntityId: EntityID<UUID>? = null
+            if (!request.projectId.isEmpty()) {
```
```suggestion
            if (request.projectId.isNotEmpty()) {
```

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -124,12 +131,33 @@ class CriterionTableRepoTest : H2DatabaseTest(arrayOf(CriterionTable, ProjectTab
+            val criterion = repo.createCriterion(request, testUserId)
 
-            assertThrows<NotFoundException> { repo.createCriterion(request, testUserId) }
+            Assertions.assertThat(criterion.tag).isEqualTo("Test Tag")
```
If you change the assertions, then you also have to change the test name.
In the test above, we create a project-criterion and in this one, a user-criterion. Let's make this clear in the test name.

domienderle replied:
Yes you are right. I made it clear.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -167,6 +195,41 @@ class CriterionTableRepoTest : H2DatabaseTest(arrayOf(CriterionTable, ProjectTab
+        @Test
+        fun `When all criteria of a specific user are requested, then the all but only the criteria created by the according user are returned`() =
+            runTest {
+                val userId = db.dbQuery {
```
Use the `createExampleUser` method from `RepositoryHelper`.

domienderle replied:
Good catch. Didn't recognize that it already exists...

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -167,6 +195,41 @@ class CriterionTableRepoTest : H2DatabaseTest(arrayOf(CriterionTable, ProjectTab
+                        .setCategory(CriterionCategory.CRITERION_CATEGORY_EXCLUSION)
+                        .build()
+                val userCriterion1 = repo.createCriterion(request, testUserId)
+                val userCriterion2 = repo.createCriterion(request, userId)
```
It would be good to add a project-criterion that was created by the test user, which then shouldn't be in the list.

domienderle replied:
Done.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -167,6 +195,41 @@ class CriterionTableRepoTest : H2DatabaseTest(arrayOf(CriterionTable, ProjectTab
+    @Nested
+    inner class GetAllUserCriteria {
+        @Test
+        fun `When all criteria of a specific user are requested, then the all but only the criteria created by the according user are returned`() =
```
```suggestion
        fun `When all criteria of a specific user are requested, then only the criteria created by the according user are returned`() =
```

Slartibartfass2 replied:
Now there appears to be a duplicated "the": "then the only the criteria"

domienderle replied:
Sorry, I don't know why it is not changed.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -91,8 +95,27 @@ class ProjectService(
-        repo.createProject(request, GrpcContext.getUserIdFromContext()).toGrpcProject()
+    override suspend fun createProject(request: GrpcProject.Create): GrpcProject {
+        val currentUser = userRepo.getUserById(GrpcContext.getUserIdFromContext())
+        val userSettings = userRepo.getUserSettings(currentUser.id)
```
Don't we have the `getAllUserCriteria` for that? When using it we don't have to fetch each criterion separately.

I just saw that we still need the user settings. However, we should still just fetch all user criteria so that we don't have to fetch each criterion separately.

domienderle replied:
That is a very good idea.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/CreateProjectTest.kt:
```diff
@@ -36,19 +38,49 @@ class CreateProjectTest : MainServiceTest() {
+    fun `When a project is correctly created and the user has default criteria, then no exception is thrown`() =
+        runTest {
+            coEvery { criterionRepoMock.getCriterionById(any()) } throws NotImplementedError()
+            coEvery { criterionRepoMock.createCriterion(any(), any()) } throws NotImplementedError()
```
Why do we mock them here and then overwrite them below?

domienderle replied:
Don't mind. I see that I got something wrong here. This is unnecessary.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/table/UserTable.kt:
```diff
@@ -28,8 +54,22 @@ object UserTable : UUIDTable("user") {
+    val snowballingType = enumeration<ProjectOuterClass.SnowballingType>(
+        "snowballing_type",
+    ).default(SNOWBALLING_TYPE_DEFAULT)
+    val reviewMaybeAllowed = bool("review_maybe_allowed").default(REVIEW_MAYBE_ALLOWED_DEFAULT)
```
Let's set the default values on the client side, i.e., our server, instead of the DB. This makes it easier to change the defaults in the future because we won't require DB migration.
```suggestion
    // User settings
    val showHotkeys = bool("show_hotkeys").clientDefault { SHOW_HOTKEYS_DEFAULT }
    val reviewMode = bool("review_mode").clientDefault { REVIEW_MODE_DEFAULT }
    val criteriaIds: Column<List<UUID>> = array<UUID>("criteria_ids").clientDefault { CRITERIA_IDS_DEFAULT }

    // Project settings defaults
    val similarityThreshold = float("similarity_threshold").clientDefault { SIMILARITY_THRESHOLD_DEFAULT }
    val decisionMatrix = binary("review_decision_matrix").clientDefault { DECISION_MATRIX_DEFAULT }
    val fetcherApis = array<String>("fetcher_apis").clientDefault { FETCHER_APIS_DEFAULT }
    val snowballingType =
        enumeration<ProjectOuterClass.SnowballingType>("snowballing_type").clientDefault { SNOWBALLING_TYPE_DEFAULT }
    val reviewMaybeAllowed = bool("review_maybe_allowed").clientDefault { REVIEW_MAYBE_ALLOWED_DEFAULT }
```

domienderle replied:
This is a good hint. I don't have any experience with database migration or that this particular solution exists. Thanks.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/table/UserTable.kt:
```diff
@@ -16,6 +30,18 @@ import java.time.OffsetDateTime
+ * [String] with a default empty list.
+ * - [snowballingType]: Represents the snowballing type associated with the user, stored as an enumeration value of
+ * [ProjectOuterClass.SnowballingType] with a default value of `SNOWBALLING_TYPE_BOTH`.
+ * - [reviewMaybeAllowed]: Indicates whether "maybe" is allowed in reviews, stored as a [Boolean] with a default value of false.
```
Let's either not mention the default values in the comment or refer to the variables because if we change a default value we will probably forget to update the comment.

domienderle replied:
Good point. I removed them.

Slartibartfass2 replied:
The first three still have the default value as plain text.

domienderle replied:
Wtf. I must have been blind. Thanks.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/table/UserTable.kt:
```diff
@@ -1,10 +1,24 @@
+val DECISION_MATRIX_DEFAULT: ByteArray = ReviewDecisionMatrix.getDefaultInstance().toByteArray()
+val FETCHER_APIS_DEFAULT = emptyList<String>()
+val SNOWBALLING_TYPE_DEFAULT = ProjectOuterClass.SnowballingType.SNOWBALLING_TYPE_BOTH
+const val REVIEW_MAYBE_ALLOWED_DEFAULT = true
```
Make those private

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/ProjectTableRepoTest.kt:
```diff
@@ -110,7 +111,8 @@ class ProjectTableRepoTest : H2DatabaseTest(arrayOf(ProjectTable, ProjectMemberT
         @Test
         fun `When a project is created, but the assigned user doesn't exist, then an exception is thrown`() = runTest {
```
In the two tests above, we use `insertTestProjectAndGetId` instead of the repo method, which is entirely wrong 😅. Can you please change that and then assert the user settings?

domienderle replied:
That is true. Why ever this was created like this. Is changed now.

Slartibartfass2 replied:
Thanks. I think this happened during a refactoring. Good that we saw and fixed this now. There should be a check that ensures that the repo method is used at least once in each test case :D

(resolve this on your own)

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -259,4 +270,23 @@ class UserService(
+        val userSettings = userRepo.getUserSettings(currentUser.id)
+        val criteria = mutableListOf<CriterionOuterClass.Criterion>()
+        for (criterionId in userSettings.criteriaIds) {
+            val criterion = criterionRepo.getCriterionById(criterionId)
```
Here again I would rather use the `getAllUserCriteria` call. Then the `criteriaIds` list in `UserSettings` is obsolete, and we probably can remove the criteriaIds from the `UserTable` all together.

Slartibartfass2 replied:
We don't have to filter the criteria according to the stored IDs since they are the same. We can just remove `criteriaIds` from the `UserTable`.

domienderle replied:
I am not sure if I get what you mean, but as far as I understand I would disagree. The `getAllUserCriteria` call returns every criterion, that was ever created by a specific user. Whereas the `userSettings.criteriaIds` (we can talk about a renaming her to somethin like `defaultCriteriaIds`) are only the ids of the criteria, the user specifically selected to be a default criterion. So the to lists are not the same. With the `getUserSettings` call we would not like to get all every created user criteria, but only the default user criteria. In this case I think the implementation is good. But tell me if I get you wrong.

Slartibartfass2 replied:
As discussed, this is fine.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/UserService.kt:
```diff
@@ -259,4 +270,23 @@ class UserService(
         return Base.Nothing.getDefaultInstance()
     }
+
+    override suspend fun getUserSettings(request: Base.Nothing): UserSettingsOuterClass.UserSettings {
```
We don't need to pass the `Nothing` request to the service layer. Just don't pass anything instead.

Slartibartfass2 reviewed (changes requested):
In conclusion, this looks very good. Great work. Some things need to be discussed further.

Now that this also includes the `getUserSettings` call, update the PR title.

domienderle commented:
Even though the pipeline fails, you can review again. I will fix the tests and the formatting issues as soon as we agree on how to handle the exceptions in the `CriterionService`.

All the changes of the current review are in the last commit.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/criterion/CreateCriterionTest.kt:
```diff
@@ -35,23 +66,117 @@ class CreateCriterionTest : MainServiceTest() {
+            "PROJECT_STATUS_ACTIVE_LOCKED",
+            "PROJECT_STATUS_DELETED",
+        ],
+    )
```
I really like that you use parameterized testing here. For gRPC enums, we can use the `GrpcEnumSourceTest`, which can exclude enum values by names since some weeks.
```suggestion
@GrpcEnumSourceTest(ProjectOuterClass.ProjectStatus::class, excludes = ["PROJECT_STATUS_ACTIVE", "PROJECT_STATUS_UNSPECIFIED"])
```

We can also include `UNSPECIFIED` because it is also non-active in this case.

domienderle replied:
That's a great suggestion, thanks.

Slartibartfass2 reviewed (changes requested):
Great work; we're nearly done here. You can already rebase. There will be two conflicts. In the first one you can just take the version of your branch, and in the second one, you just have to change the mapping of the switch expression.

domienderle commented:
I have rebased the commits of the last round into the previous commits, thus the last commit now again contains the changes of this round.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/MainService.kt:
These changes should be undone. Probably an error during rebasing.

domienderle replied:
Yes, it was my mistake during rebasing. For some reason I thought, that I don't need it anymore, but I did. The problem is, when I rebase again I don't get the newest version of the file from develop back. I just copied the file now, but it is still in the newest review commit.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/service/ProjectService.kt:
```diff
@@ -91,8 +95,27 @@ class ProjectService(
+                .setName(criterion.name)
+                .setDescription(criterion.description)
+                .setCategory(criterion.category)
+                .setProjectId(project.id ?: "")
```
Do we require this null fallback (elvis operator)? The following should be just fine.
```suggestion
                .setProjectId(project.id)
```

domienderle replied:
If we don't add it, we get a linter error. So either we let it as it is or we add a suppress linter annotation. I don't care.

Slartibartfass2 replied:
Ah, I didn't check the linter. This is because the gRPC code is in Java, and therefore in Kotlin every property is assumed to be nullable.
We don't need to convert the project to a gRPC object until the end of this method, so we could just use the DTO for this and then call `.setProjectId(project.id.toString())` and at the end `return project.toGrpcProject()`.

domienderle replied:
You are right. Could have figured this out myself... Thank you.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/auth/DummyUser.kt:
```diff
@@ -20,4 +21,14 @@ object DummyUser {
     var status: UserOuterClass.UserStatus = UserOuterClass.UserStatus.USER_STATUS_ACTIVE
+
+    // user settings
+    var isShowHotkeysEnabled: Boolean = true
```
This seems a bit "doppeltgemoppelt", what about `areHotkeysShown`, `areHotkeysVisible`, or `hotkeyVisibiltyEnabled`?

Furthermore, I would keep this name consistent, i.e., in `DummyUser`, `UserSettings`, and `UserTable` (including the column name and default value)

domienderle replied:
You are right. Sounds better.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/table/UserTable.kt:
```diff
@@ -28,8 +51,20 @@ object UserTable : UUIDTable("user") {
-    // Metadata
+    // User settings
+    val showHotkeys = bool("show_hotkeys").clientDefault { SHOW_HOTKEYS_DEFAULT }
+    val reviewMode = bool("review_mode").clientDefault { REVIEW_MODE_DEFAULT }
```
Let's keep this name consistent with the user settings variable.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/association/ProjectMemberTableRepoTest.kt:
```diff
@@ -17,6 +20,8 @@ import snowballr.ProjectOuterClass
 import java.util.UUID
 
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
These can be removed since #231.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepoTest.kt:
```diff
@@ -11,17 +13,22 @@ import org.junit.jupiter.api.assertThrows
 import java.util.UUID
 
+@ExperimentalCoroutinesApi
+@DelicateCoroutinesApi
```
These can be removed since #231.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/repository/CriterionTableRepo.kt:
```diff
@@ -122,4 +140,18 @@ class CriterionTableRepo(
+            .map { it.toCriterion() }
+    }
+
+    override suspend fun getAllUserCriteria(userId: UUID): List<Criterion> = db.dbQuery {
```
Now this is unused. Should we keep it?

domienderle replied:
I would propose to keep it. Since, we need this call at latest, when we want to give the user an overview over all his criteria to select the default criteria. Could take a while until we implement this in the frontend, but since it is already implemented here, it saves work later and would keep it.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/CreateProjectTest.kt:
```diff
@@ -12,15 +12,17 @@ import se.uulm.snowballr.backend.TestSpecificException
     private fun getExampleRequest() = ProjectOuterClass.Project.Create.getDefaultInstance()
 
     @BeforeEach
     fun setupTest() {
```
We don't mock or test the criterion calls `getCriteriaByIds` and `createCriterion`. We can test the case that `getCriteriaByIds` returns a list of IDs and then can verify whether `createCriterion` is called with the IDs as parameters. Another test case would be that an empty list is returned and `createCriterion` isn't called at all. For the rest of the test cases, they don't have to do anything, but we should add these two cases.

To verify that the method was called with certain parameters, we can use:
```kotlin
val slots = mutableListOf<UUID>()
coEvery { criterionRepoMock.getCriterionById(capture(slots)) }
// ... assert that slots match the list of IDs
```

To verify that the method was not called at all, we can use:
```kotlin
verify(exactly = 0) { criterionRepoMock.createCriterion(any()) }
```

domienderle replied:
Oh shit good call. I completely forgot to include them. Thanks.

Slartibartfass2 reviewed (changes requested):
The tests of the newly added repo method are well written; good job.

Only some minor things and two missing test cases. You can rebase already.

Slartibartfass2 commented on src/test/kotlin/se/uulm/snowballr/backend/service/project/CreateProjectTest.kt:
```diff
@@ -25,19 +26,50 @@ class CreateProjectTest : MainServiceTest() {
+            every { GrpcContext.getUserIdFromContext() } returns user.id
+            coEvery { userRepoMock.getUserById(user.id) } returns user
+            coEvery { userRepoMock.getUserSettings(user.id) } returns userSettings
+            coEvery { criterionRepoMock.getCriteriaByIds(slots) } returns listOf(criterion)
```
This is not how slots are used.
Pass an empty mutable list to `coEvery { criterionRepoMock.getCriteriaByIds(capture(slots)) }`
and then assert after the service call has been made the content of the list.

Refer to [the Mockk documentation](https://mockk.io/#capturing) for more information.

domienderle replied:
Ok. I hope I implemented it the correct way now.

Slartibartfass2 reviewed (changes requested):
Only these two small things.

Slartibartfass2 commented on src/main/kotlin/se/uulm/snowballr/backend/auth/DummyUser.kt:
```diff
@@ -20,4 +21,14 @@ object DummyUser {
     var status: UserOuterClass.UserStatus = UserOuterClass.UserStatus.USER_STATUS_ACTIVE
+
+    // user settings
+    var areHotkeysShown: Boolean = true
+    var isReviewModeEnabled: Boolean = false
+    var criteriaIds: List<UUID> = emptyList()
+    var similarityThreshold: Float = 0F
+    var decisionMatrix: ByteArray = ProjectOuterClass.ReviewDecisionMatrix.getDefaultInstance().toByteArray()
+    var fetcherApis: List<String> = emptyList()
+    var snowballingType: ProjectOuterClass.SnowballingType = ProjectOuterClass.SnowballingType.SNOWBALLING_TYPE_BOTH
+    var reviewMaybeAllowed: Boolean = true
```
Another small thing. There is a test for the dummy user. Just to have the coverage. Add the assertions for these properties to `DummyUserTest`.

domienderle committed:
chore: include review changes

Slartibartfass2 reviewed (approved):
Again, great work! This took some time, but now this is another great addition.
